File: findAllNumberDisappearedInAnArray.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    vector<int> findDisappearedNumbers(vector<int> &nums)
    {
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums.at(abs(nums.at(i)) - 1) > 0)
            {
                nums.at(abs(nums.at(i)) - 1) *= (-1);
            }
        }
        vector<int> ans;
        for (auto x = 0; x < nums.size(); x++)
        {
            if (nums.at(x) > 0)
            {
                ans.push_back(x + 1);
            }
        }
        return ans;
    }
};

int main()
{

    Solution s = Solution();

    vector<int> vc = {4, 3, 2, 7, 8, 2, 3, 1};
    s.findDisappearedNumbers(vc);

    return 0;
}

File: FindAllNumberDisappearedInAnArray.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;

public class FindAllNumberDisappearedInAnArray {

    public static void main(String[] args) {

        Solution s = new Solution();
        int nums[] = { 4, 3, 2, 7, 8, 2, 3, 1 };
        s.findDisappearedNumbers(nums);
    }
}

class Solution {
    // using extrspace
    public List<Integer> findDisappearedNumbers(int[] nums) {

        HashSet<Integer> ht = new HashSet<>();
        int n = nums.length;
        for (int i = 1; i <= n; i++) {

            ht.add(i);
        }
        for (int i = 0; i < n; i++) {
            if (ht.contains(nums[i])) {
                ht.remove(nums[i]);
            }
        }
        var it = ht.iterator();
        List<Integer> ans = new ArrayList<>();
        while (it.hasNext()) {

            ans.add(it.next());
        }

        return ans;

    }
}